src/sema/namespace.rs

```diff
@@ -490,6 +507,106 @@ impl Namespace {
+        // or import symbol), then we should search both the current contract and global scope.
+        if path.identifiers.len() == 1 {
+            let id = &path.identifiers[0];
+
+            // If we're in a contract, error can be defined in current contract or its bases
+            if let Some(contract_no) = contract_no {
+                for contract_no in self.contract_bases(contract_no).into_iter().rev() {
+                    let file_no = self.contracts[contract_no].loc.file_no();
+
+                    match self.variable_symbols.get(&(
+                        file_no,
+                        Some(contract_no),
+                        id.name.to_owned(),
+                    )) {
+                        None => (),
+                        Some(Symbol::Error(_, error_no)) => {
+                            return Ok(*error_no);
+                        }
+                        sym => {
+                            let error = Namespace::wrong_symbol(sym, id);
+
+                            diagnostics.push(error);
+
+                            return Err(());
+                        }
+                    }
+
+                    if let Some(sym) =
+                        self.function_symbols
+                            .get(&(file_no, Some(contract_no), id.name.to_owned()))
+                    {
+                        let error = Namespace::wrong_symbol(Some(sym), id);
+
+                        diagnostics.push(error);
+
+                        return Err(());
+                    }
+                }
+            }
+
+            if let Some(sym) = self
+                .function_symbols
+                .get(&(file_no, None, id.name.to_owned()))
+            {
+                let error = Namespace::wrong_symbol(Some(sym), id);
+
+                diagnostics.push(error);
+
+                return Err(());
+            }
+
+            return match self
+                .variable_symbols
+                .get(&(file_no, None, id.name.to_owned()))
+            {
+                None => {
+                    diagnostics.push(Diagnostic::decl_error(
+                        id.loc,
+                        format!("error '{}' not found", id.name),
+                    ));
+                    Err(())
+                }
+                Some(Symbol::Error(_, error_no)) => Ok(*error_no),
+                sym => {
+                    let error = Namespace::wrong_symbol(sym, id);
+
+                    diagnostics.push(error);
+
+                    Err(())
+                }
+            };
+        }
+
+        let (id, namespace) = path
+            .identifiers
+            .split_last()
+            .map(|(id, namespace)| (id, namespace.iter().collect()))
+            .unwrap();
+
+        let s = self.resolve_namespace(namespace, file_no, contract_no, id, diagnostics)?;
+
+        if let Some(Symbol::Error(_, error_no)) = s {
```

Comment:
On a quick glance it looks like [ok_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) could be used here

Reply:
`Option.ok_or_else()` transforms `Some(x)` to `Ok(x)` or else calls a closure. In the `Ok()` cases we still want to extract `error_no` from `Symbol::Error(_, error_no)` and only match `Symbol::Error`. I am not sure how `ok_or_else()` can be used here.

Did not know about this function, looks very useful!